src/Build/Utilities/EngineFileUtilities.cs

```diff
@@ -152,8 +219,193 @@ FileMatcher fileMatcher
+                                if (((!excludeFileSpecIsUnusable) && (!disableExcludeDriveEnumerationWarning)) ||
+                                    (includeLocation == null))
+                                {
+                                    targetLoggingContext.LogWarning(
+                                            DriveEnumeratingWildcardMessageResourceName,
+                                            EscapingUtilities.UnescapeAll(fileSpec),
+                                            XMakeAttributes.exclude,
+                                            XMakeElements.itemGroup);
+                                }
+
+                                // Both conditions are necessary to reach for both GetFileListEscaped calls
+                                // and skip for the GetFileListUnescaped call when the wildcarded Include attribute
+                                // results in drive enumeration.
+                                else if (excludeFileSpecIsUnusable && (includeLocation != null))
+                                {
+                                    targetLoggingContext.LogWarning(
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(fileSpec),
+                                        XMakeAttributes.include,
+                                        XMakeElements.itemGroup);
+                                }
+
+                                break;
+
+                            // Logging mechanism received from Evaluator.
+                            case ILoggingService loggingService:
+                                if (buildEventContext != null)
+                                {
+                                    loggingService.LogWarning(
+                                        buildEventContext,
+                                        string.Empty,
+                                        new BuildEventFileInfo(buildEventFileInfoFullPath),
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(filespecUnescaped),
+                                        XMakeAttributes.include,
+                                        XMakeElements.itemGroup);
+                                }
+
+                                break;
+
+                            // Logging mechanism received from Evaluator and LazyItemEvaluator.IncludeOperation.
+                            case EvaluationLoggingContext evaluationLoggingContext:
+                                if (importLocation != null)
+                                {
+                                    evaluationLoggingContext.LogWarning(
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(filespecUnescaped),
```

Comment:
Is this unescaping of an already unescaped filespec correct?

Reply:
Just double checked FileMatcher for the `excludeFileSpec` case as well - it's not correct, neither `filespecUnescaped` nor `excludeFileSpec` should be unescaped, so that should be removed.